File: dodian-backend/dodian-server/src/main/java/net/dodian/uber/game/model/ShopHandler.java
package net.dodian.uber.game.model;

import net.dodian.utilities.Utils;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

public class ShopHandler {
    public static int MaxShops = 101; // 1 more because we don't use [0] !
    public static int MaxShopItems = 40;
    public static int MaxShowDelay = 100;
    public static int TotalShops = 0;
    public static int[][] ShopItems = new int[MaxShops][MaxShopItems];
    public static int[][] ShopItemsN = new int[MaxShops][MaxShopItems];
    public static int[] ShopItemsDelay = new int[MaxShops];
    public static int[][] ShopItemsSN = new int[MaxShops][MaxShopItems];
    public static int[] ShopItemsStandard = new int[MaxShops];
    public static String[] ShopName = new String[MaxShops];
    public static int[] ShopSModifier = new int[MaxShops];
    public static int[] ShopBModifier = new int[MaxShops];

    public ShopHandler() {
        for (int i = 0; i < MaxShops; i++) {
            for (int j = 0; j < MaxShopItems; j++) {
                ResetItem(i, j);
                ShopItemsSN[i][j] = 0; // Special resetting, only at begin !
            }
            ShopItemsStandard[i] = 0; // Special resetting, only at begin !
            ShopSModifier[i] = 0;
            ShopBModifier[i] = 0;
            ShopName[i] = "";
        }
        TotalShops = 0;
        loadShops("./data/shops.cfg");
    }

    public void DiscountItem(int ShopID, int ArrayID) {
        ShopItemsN[ShopID][ArrayID] -= 1;
        if (ShopItemsN[ShopID][ArrayID] <= 0 && ArrayID >= ShopHandler.ShopItemsStandard[ShopID]) {
            ShopItemsN[ShopID][ArrayID] = 0;
            ResetItem(ShopID, ArrayID);
        }
    }

    public void ResetItem(int ShopID, int ArrayID) {
        ShopItems[ShopID][ArrayID] = 0;
        ShopItemsN[ShopID][ArrayID] = 0;
        ShopItemsDelay[ShopID] = 0;
    }

    public static void resetAnItem(int ShopID, int ArrayID) {
        ShopItems[ShopID][ArrayID] = -1;
        ShopItemsN[ShopID][ArrayID] = 0;
        ShopItemsDelay[ShopID] = 0;
    }

    public static boolean findDefaultItem(int shopId, int id) {
        for(int i = 0; i < ShopItemsStandard[shopId]; i++)
            if(ShopItems[shopId][i] -1 == id)
                return true;
        return false;
    }

    @SuppressWarnings("resource")
    public boolean loadShops(String FileName) {
        String line = "";
        String token = "";
        String token2 = "";
        String token2_2 = "";
        String[] token3 = new String[(MaxShopItems * 2) + 4];
        boolean EndOfFile = false;
        BufferedReader characterfile = null;
        try {
            characterfile = new BufferedReader(new FileReader("./" + FileName));
        } catch (FileNotFoundException fileex) {
            Utils.println(FileName + ": file not found.");
            return false;
        }
        try {
            line = characterfile.readLine();
        } catch (IOException ioexception) {
            Utils.println(FileName + ": error loading file.");
            return false;
        }
        while (EndOfFile == false && line != null) {
            line = line.trim();
            int spot = line.indexOf("=");
            if (spot > -1) {
                token = line.substring(0, spot);
                token = token.trim();
                token2 = line.substring(spot + 1);
                token2 = token2.trim();
                token2_2 = token2.replaceAll("\t\t", "\t");
                token2_2 = token2_2.replaceAll("\t\t", "\t");
                token2_2 = token2_2.replaceAll("\t\t", "\t");
                token2_2 = token2_2.replaceAll("\t\t", "\t");
                token2_2 = token2_2.replaceAll("\t\t", "\t");
                token3 = token2_2.split("\t");
                if (token.equals("shop")) {
                    int ShopID = Integer.parseInt(token3[0]);
                    ShopName[ShopID] = token3[1].replaceAll("_", " ");
                    ShopSModifier[ShopID] = Integer.parseInt(token3[2]);
                    ShopBModifier[ShopID] = Integer.parseInt(token3[3]);
                    for (int i = 0; i < ((token3.length - 4) / 2); i++) {
                        if (token3[(4 + (i * 2))] != null) {
                            ShopItems[ShopID][i] = (Integer.parseInt(token3[(4 + (i * 2))]) + 1);
                            ShopItemsN[ShopID][i] = Integer.parseInt(token3[(5 + (i * 2))]);
                            ShopItemsSN[ShopID][i] = Integer.parseInt(token3[(5 + (i * 2))]);
                            ShopItemsStandard[ShopID]++;
                        } else {
                            break;
                        }
                    }
                    TotalShops++;
                }
            } else {
                if (line.equals("[ENDOFSHOPLIST]")) {
                    try {
                        characterfile.close();
                    } catch (IOException ioexception) {
                    }
                    return true;
                }
            }
            try {
                line = characterfile.readLine();
            } catch (IOException ioexception1) {
                EndOfFile = true;
            }
        }
        try {
            characterfile.close();
        } catch (IOException ioexception) {
        }
        return false;
    }
}

File: game-server/src/main/kotlin/net/dodian/uber/game/model/ShopHandler.kt
package net.dodian.uber.game.model

import java.io.BufferedReader
import java.io.FileNotFoundException
import java.io.FileReader
import java.io.IOException
import net.dodian.utilities.Utils

class ShopHandler {
    init {
        for (i in 0 until MaxShops) {
            for (j in 0 until MaxShopItems) {
                ResetItem(i, j)
                ShopItemsSN[i][j] = 0
            }
            ShopItemsStandard[i] = 0
            ShopSModifier[i] = 0
            ShopBModifier[i] = 0
            ShopName[i] = ""
        }
        TotalShops = 0
        loadShops("./data/shops.cfg")
    }

    fun DiscountItem(ShopID: Int, ArrayID: Int) {
        ShopItemsN[ShopID][ArrayID] -= 1
        if (ShopItemsN[ShopID][ArrayID] <= 0 && ArrayID >= ShopItemsStandard[ShopID]) {
            ShopItemsN[ShopID][ArrayID] = 0
            ResetItem(ShopID, ArrayID)
        }
    }

    fun ResetItem(ShopID: Int, ArrayID: Int) {
        ShopItems[ShopID][ArrayID] = 0
        ShopItemsN[ShopID][ArrayID] = 0
        ShopItemsDelay[ShopID] = 0
    }

    @Suppress("resource")
    fun loadShops(FileName: String): Boolean {
        var line: String
        val token3 = arrayOfNulls<String>((MaxShopItems * 2) + 4)

        val characterfile: BufferedReader =
            try {
                BufferedReader(FileReader("./$FileName"))
            } catch (_: FileNotFoundException) {
                Utils.println("$FileName: file not found.")
                return false
            }

        try {
            line = characterfile.readLine() ?: return false
        } catch (_: IOException) {
            Utils.println("$FileName: error loading file.")
            return false
        }

        var endOfFile = false
        while (!endOfFile && line.isNotEmpty()) {
            val trimmed = line.trim()
            val spot = trimmed.indexOf('=')
            if (spot > -1) {
                val token = trimmed.substring(0, spot).trim()
                val token2 = trimmed.substring(spot + 1).trim()
                val token2_2 = token2.replace("\t\t", "\t").replace("\t\t", "\t").replace("\t\t", "\t").replace("\t\t", "\t").replace("\t\t", "\t")
                val parsed = token2_2.split("\t")
                for (i in parsed.indices) {
                    token3[i] = parsed[i]
                }
                if (token == "shop") {
                    val ShopID = token3[0]!!.toInt()
                    ShopName[ShopID] = token3[1]!!.replace("_", " ")
                    ShopSModifier[ShopID] = token3[2]!!.toInt()
                    ShopBModifier[ShopID] = token3[3]!!.toInt()
                    for (i in 0 until ((parsed.size - 4) / 2)) {
                        val itemValue = token3[4 + (i * 2)] ?: break
                        ShopItems[ShopID][i] = itemValue.toInt() + 1
                        ShopItemsN[ShopID][i] = token3[5 + (i * 2)]!!.toInt()
                        ShopItemsSN[ShopID][i] = token3[5 + (i * 2)]!!.toInt()
                        ShopItemsStandard[ShopID]++
                    }
                    TotalShops++
                }
            } else if (trimmed == "[ENDOFSHOPLIST]") {
                try {
                    characterfile.close()
                } catch (_: IOException) {
                }
                return true
            }

            line =
                try {
                    characterfile.readLine() ?: break
                } catch (_: IOException) {
                    endOfFile = true
                    ""
                }
        }

        try {
            characterfile.close()
        } catch (_: IOException) {
        }
        return false
    }

    companion object {
        @JvmField var MaxShops: Int = 101
        @JvmField var MaxShopItems: Int = 40
        @JvmField var MaxShowDelay: Int = 100
        @JvmField var TotalShops: Int = 0

        @JvmField var ShopItems: Array<IntArray> = Array(MaxShops) { IntArray(MaxShopItems) }
        @JvmField var ShopItemsN: Array<IntArray> = Array(MaxShops) { IntArray(MaxShopItems) }
        @JvmField var ShopItemsDelay: IntArray = IntArray(MaxShops)
        @JvmField var ShopItemsSN: Array<IntArray> = Array(MaxShops) { IntArray(MaxShopItems) }
        @JvmField var ShopItemsStandard: IntArray = IntArray(MaxShops)
        @JvmField var ShopName: Array<String> = Array(MaxShops) { "" }
        @JvmField var ShopSModifier: IntArray = IntArray(MaxShops)
        @JvmField var ShopBModifier: IntArray = IntArray(MaxShops)

        @JvmStatic
        fun resetAnItem(ShopID: Int, ArrayID: Int) {
            ShopItems[ShopID][ArrayID] = -1
            ShopItemsN[ShopID][ArrayID] = 0
            ShopItemsDelay[ShopID] = 0
        }

        @JvmStatic
        fun findDefaultItem(shopId: Int, id: Int): Boolean {
            for (i in 0 until ShopItemsStandard[shopId]) {
                if (ShopItems[shopId][i] - 1 == id) {
                    return true
                }
            }
            return false
        }
    }
}
